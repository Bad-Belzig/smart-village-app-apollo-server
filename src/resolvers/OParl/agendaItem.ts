import {
  AgendaItem,
  Consultation,
  File,
  IAgendaItem,
  Meeting,
} from '../../models';

export const agendaItemResolvers = {
  Query: {
    oParlAgendaItems: () => AgendaItem.find(),
  },
  OParlAgendaItem: {
    meeting: (args: IAgendaItem) =>
      Meeting.findOne({ externalId: args.meeting }),
    consultation: (args: IAgendaItem) =>
      Consultation.findOne({ externalId: args.consultation }),
    resolutionFile: (args: IAgendaItem) =>
      File.findOne({ externalId: args.resolutionFile }),
    auxiliaryFile: (args: IAgendaItem) =>
      args.auxiliaryFile?.map((value) => File.findOne({ externalId: value })),
  },
};
