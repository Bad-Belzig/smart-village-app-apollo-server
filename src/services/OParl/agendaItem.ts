import { AgendaItem } from '../../models';
import { parseAgendaItem } from '../../parser';
import { getJson, updateOrCreateEntry } from './importHelpers';
import { ImportQueue, ImportQueueEntry, ImportType } from './ImportTypes';

export const importAgendaItem = async (value: unknown, queue?: ImportQueue) => {
  const json = await getJson(value);

  if (!json) return;

  const addToQueue: [ImportQueueEntry | ImportQueueEntry[], ImportType][] = [];

  if (json.consultation) {
    addToQueue.push([json.consultation, ImportType.Consultation]);
  }

  if (json.meeting) {
    addToQueue.push([json.meeting, ImportType.Meeting]);
  }

  if (json.resolutionFile) {
    addToQueue.push([json.resolutionFile, ImportType.File]);
  }

  if (json.auxiliaryFile) {
    addToQueue.push([json.auxiliaryFile, ImportType.File]);
  }

  try {
    return await updateOrCreateEntry(
      json,
      parseAgendaItem,
      AgendaItem,
      addToQueue,
      queue,
    );
  } catch (e) {
    console.log(`Error while importing agenda item from ${value}:`, e);
  }
};
