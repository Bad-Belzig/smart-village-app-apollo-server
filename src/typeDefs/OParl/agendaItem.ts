import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const agendaItemGQL = gql`
  extend type Query {
    oParlAgendaItems(
      externalIds: [String!]
      keyword: [String!]
    ): [OParlAgendaItem!]
  }

  type OParlAgendaItem {
    auxiliaryFile(offset: Int, pageSize: Int): [OParlFile!]
    consultation: OParlConsultation
    end: Date
    meeting: OParlMeeting
    name: String
    number: String
    order: Int
    public: Boolean
    resolutionFile: OParlFile
    resolutionText: String
    result: String
    start: Date
    ${oParlBase}
  }
`;
