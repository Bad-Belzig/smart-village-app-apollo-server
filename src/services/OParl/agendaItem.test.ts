import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../testHelpers';
import { importAgendaItem } from './agendaItem';

const helperFuncs = setupTestDataBase('test_agenda_item_import');

// TODO: change test url to something within our control, that uses all fields
const testUrl =
  'https://www.muenchen-transparent.de/oparl/v1.0/agendaitem/48823';

describe('importing a agenda item', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importAgendaItem, testUrl));

  it(
    'import returns undefined on faulty url',
    invalidUrlTest(importAgendaItem),
  );

  afterAll(helperFuncs.afterAll);
});
