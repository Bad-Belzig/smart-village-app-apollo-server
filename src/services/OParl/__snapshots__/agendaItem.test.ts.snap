// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a agenda item imports properly for 1.0 1`] = `
Object {
  "__v": 0,
  "auxiliaryFile": Array [],
  "consultation": "https://www.muenchen-transparent.de/oparl/v1.0/consultation/48823",
  "externalId": "https://www.muenchen-transparent.de/oparl/v1.0/agendaitem/48823",
  "keyword": Array [],
  "name": "Tagesordnung",
  "number": "1",
  "resolutionText": "",
  "type": "https://schema.oparl.org/1.0/AgendaItem",
}
`;

exports[`importing a agenda item imports properly for 1.1 1`] = `
Object {
  "__v": 0,
  "auxiliaryFile": Array [],
  "created": 2015-09-09T11:55:58.000Z,
  "externalId": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/agendaitem/1355",
  "keyword": Array [],
  "meeting": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/meeting/485",
  "modified": 2015-10-08T09:15:47.000Z,
  "name": "Ã–ffentliche Sitzung",
  "number": "I.",
  "order": 1,
  "type": "https://schema.oparl.org/1.1/AgendaItem",
}
`;
